homeassistant/util/secrets.py

```diff
@@ -0,0 +1,133 @@
+"""Module to help keep passwords and other secrets."""
+
+import os
+import logging
+import collections
+from enum import Enum
+from homeassistant.config import load_yaml_config_file
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.const import EVENT_HOMEASSISTANT_START
+try:
+    import keyring  # pylint: disable=wrong-import-order,wrong-import-position
+except ImportError:
+    keyring = None
+
+# Contains secrets resolved by get_secret (from config files & components)
+# History can be used to edit secrets through the frontend after HA started.
+HISTORY = {}
+"""HISTORY will be used to store secrets extracted from the configuration files
+and make it available for editing later."""  # pylint: disable=W0105
+SECRET_PLACEHOLDER = '(-)'
+SECRET_DICT = None
+_LOGGER = logging.getLogger(__name__)
+_LOGGER.setLevel(logging.DEBUG)
+
+
+class SSource(Enum):
+    """Source for secrets."""
+
+    yaml = 1
+    keyring = 2
+
+
+def get_secret(namespace, username):
+    """Retrieve a secret from the secrets.yaml or keyring.
+
+    The full secret path will be: namespace/username
+
+    Once retrieved, the source will be saved in the HISTORY
+
+    Depending on availability of the system and existance of the secret's path,
+    the secret will be extracted in the following order:
+      (1) secrets.yaml: Located in the configuration.yaml directory
+          Format: /path/to/username: secret
+      (2) keyring: http://pythonhosted.org/keyring/
+    """
+    fullpath = (str(namespace) + '/' + str(username)).replace(' ', '_')
+
+    global HISTORY  # pylint: disable=global-variable-not-assigned
```

Comment:
Ok, will remove, guess since its mutable and initialized elsewhere I only ever read from the global var